## 1. Evaluation initiale sur le dataset Adult

### 1.1 Chargement et exécution des modèles sur le dataset Adult
Elle charge les données Adult et exécute les trois modèles (XGBoost, BRCG, RIPPER) pour comparer leurs performances.

In [3]:
import import_ipynb  
import brcg_model
import adultdata_preparation_evaluat
import xgboost_rule_extraction
import explainer
import pandas as pd


# Charger les données
x_train, x_test, y_train, y_test, mapping = adultdata_preparation_evaluat.prepare_adult_dataset()
print("Données chargées depuis le notebook principal") 

# Nettoyer les données                          
x_train = x_train.drop(columns=['label'], errors='ignore')
x_test = x_test.drop(columns=['label'], errors='ignore')

# Exécuter le notebook XGBoost - cela va définir train_xgboost dans l'espace de noms
%run ./xgboost_rule_extraction.ipynb

# Vérifier que train_xgboost est maintenant disponible
print("train_xgboost disponible:", 'train_xgboost' in dir())

if 'train_xgboost' in dir():
    # Utiliser XGBoost
    
    x_test = x_test.drop(columns=['label'], errors='ignore')
    xgb_model, y_pred_xgb, xgb_metrics = xgboost_rule_extraction.train_xgboost(
        x_train, y_train, x_test, y_test
    )
    print("Métriques XGBoost:", xgb_metrics)
else:
    print("ERREUR: train_xgboost n'a pas été chargé correctement")
    print("Contenu de l'espace de noms:", [v for v in dir() if 'train' in v])

# BRCG
brcg_model_, fb_brcg, columns_brcg, train_time_brcg, brcg_metrics, brcg_rules = brcg_model.train_brcg(
    x_train, y_train, x_test, y_test
)
print("Métriques BRCG:", brcg_metrics)

# RIPPER
from ripper_model import train_ripper
ripper_model_, y_pred_ripper, ripper_metrics, ripper_rules = train_ripper(
    x_train, y_train, x_test, y_test
)
print("Métriques RIPPER:", ripper_metrics)

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.2181
Iteration: 2, Objective: 0.1897
Iteration: 3, Objective: 0.1854
Iteration: 4, Objective: 0.1843
Iteration: 5, Objective: 0.1843
Iteration: 6, Objective: 0.1843
Résultats BRCG
Accuracy: 0.8189774297558728
F1-score: 0.8026137894465102
Precision: 0.8063169426792735
Recall: 0.8189774297558728
Impossible d'afficher les règles


100%|██████████| 14/14 [00:00<00:00, 16.77it/s]


Total Support: 23.48% (6116)
After removing duplicates:
Unique Coverage: 21.88% (5698)


100%|██████████| 723/723 [00:00<00:00, 3600.70it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using Majority class to predict non-covered cases.
F1-score (if using entire dataset): 0.7775


100%|██████████| 723/723 [00:00<00:00, 3989.28it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using model to predict non-covered cases.
F1-score (if using entire dataset): 0.8689
Données chargées depuis le notebook principal


100%|██████████| 14/14 [00:00<00:00, 18.89it/s]


Total Support: 23.48% (6116)
After removing duplicates:
Unique Coverage: 21.88% (5698)


100%|██████████| 723/723 [00:00<00:00, 3951.97it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using Majority class to predict non-covered cases.
F1-score (if using entire dataset): 0.7775


100%|██████████| 723/723 [00:00<00:00, 3643.01it/s]


Test set coverage: 18.26%
Test coverage F1-score: 0.98
--------------------------------------
Total Skipped: 5324 (81.74%)
 - Due to Conflicts: 0 (0.0%)
 - Due to No Rule Coverage: 5324 (81.74%)
--------------------------------------
Using model to predict non-covered cases.
F1-score (if using entire dataset): 0.8689
train_xgboost disponible: True
Métriques XGBoost: {'accuracy': 0.8753262705358513, 'balanced_accuracy': 0.7969251078262138, 'precision': 0.7990543735224587, 'recall': 0.645448758752387, 'f1': 0.7140845070422536, 'accuracy_train': 0.8748080466830467, 'f1_train': 0.7087092451987494, 'training_time': 0.4665820598602295}
Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.2181
Iteration: 2, Objective: 0.1897
Iteration: 3, Objective: 0.1854
Iteration: 4, Objective: 0.1843
Iteration: 5, Objective: 0.1843
Iteration: 6, Objective: 0.1843
Métriques BRCG: {'accuracy': 0.8189774297558728, 'f1': 0.8026137894465102, 'pr

### 1.2 Comparaison des performances des modèles

In [4]:
print("\n===== Comparaison des modèles =====")

print(f"XGBoost  -> Accuracy: {xgb_metrics['accuracy']:.4f} | F1-score: {xgb_metrics['f1']:.4f}")
print(f"BRCG     -> Accuracy: {brcg_metrics['accuracy']:.4f} | F1-score: {brcg_metrics['f1']:.4f}")
print(f"RIPPER   -> Accuracy: {ripper_metrics['accuracy']:.4f} | F1-score: {ripper_metrics['f1']:.4f}")


===== Comparaison des modèles =====
XGBoost  -> Accuracy: 0.8753 | F1-score: 0.7141
BRCG     -> Accuracy: 0.8190 | F1-score: 0.8026
RIPPER   -> Accuracy: 0.8188 | F1-score: 0.8009


--> XGBoost surpasse en accuracy mais souffre d'un déséquilibre de classes (F1-score plus faible)

--> RIPPER offre le meilleur compromis avec un F1-score élevé et une bonne accuracy

--> BRCG est très proche de RIPPER en performance

## 2. Évaluation multicritères sur les 5 datasets (BRCG RIPPER)

### 2.1 Import des bibliothèques pour l'évaluation 

In [17]:
import warnings
warnings.filterwarnings("ignore")
import os
import gc
import time
import psutil
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import traceback
from tqdm import tqdm

from brcg_model import train_brcg, evaluate_brcg
from ripper_model import train_ripper, evaluate_ripper
from cmr_model import run_cmr

import import_ipynb
import ripper_model
import brcg_model

os.makedirs("results", exist_ok=True)

In [ ]:
# Mesure le temps d'exécution et le pic de mémoire RAM 

import threading

def run_with_peak_ram(func, *args, interval=0.05, **kwargs):
    process = psutil.Process()
    baseline_rss = process.memory_info().rss
    peak_rss = {"value": baseline_rss}
    stop_flag = {"stop": False}
    result_holder = {}

    # Fonction de monitoring de la mémoire
    def monitor():
        while not stop_flag["stop"]:
            try:
                rss = process.memory_info().rss
                if rss > peak_rss["value"]:
                    peak_rss["value"] = rss
            except:
                pass
            time.sleep(interval)

    t = threading.Thread(target=monitor, daemon=True)

    start = time.time()
    t.start()

    try:
        result_holder["result"] = func(*args, **kwargs)
    finally:
        stop_flag["stop"] = True
        t.join()

    elapsed = time.time() - start

    added_ram_mb = max(0.0, (peak_rss["value"] - baseline_rss) / 1024**2)
    peak_ram_mb = peak_rss["value"] / 1024**2

    return result_holder["result"], elapsed, added_ram_mb, peak_ram_mb

### 2.2 Évaluation avec 1 iterations

In [32]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

    # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None

    return results

datasets = {
    "KDD99": "../Datasets/KDD99/",
    "BIG15": "../Datasets/BIG15/",
    "UNSW-NB15": "../Datasets/UNSW-NB15/processed/",
    "DoH20": "../Datasets/DoH20/"
}

results_all_iterations = {ds: [] for ds in datasets.keys()}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        if name == "UNSW-NB15":
            train = pd.read_csv(path + "le_train.csv")
            test = pd.read_csv(path + "le_test.csv")

            X_train = train.drop(["id", "attack_cat", "label"], axis=1)
            y_train = train["label"]

            X_test = test.drop(["id", "attack_cat", "label"], axis=1)
            y_test = test["label"]
        else:
            X_train = pd.read_csv(path + "X_train.csv")
            X_test = pd.read_csv(path + "X_test.csv")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=None,
                dataset_name=name,
                iteration=it+1
            )
            results_all_iterations[name].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = [name, method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations.csv", index=False)
        
        if name == "UNSW-NB15":
            del X_train, X_test, y_train, y_test, train, test
        else:
            del X_train, X_test, y_train, y_test
        gc.collect()


    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds in datasets.keys():
    runs = results_all_iterations.get(ds, [])

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = [ds, method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif.csv", index=False)

if os.path.exists("results/progress_iterations.csv"):
    df_progress = pd.read_csv("results/progress_iterations.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}.png")
        plt.close()

print(df_results)

  0%|          | 0/4 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 25%|██▌       | 1/4 [08:07<24:21, 487.15s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 50%|█████     | 2/4 [19:28<20:02, 601.22s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 75%|███████▌  | 3/4 [32:11<11:15, 675.45s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 4/4 [1:53:27<00:00, 1701.79s/it]


      Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
3       BIG15    BRCG  0.981789  0.975395             NaN          0.981736   
4       BIG15  RIPPER  0.982605  0.982682             NaN          0.982578   
5       BIG15     CMR  0.994849  0.993774           95.15          1.000000   
9       DoH20    BRCG  0.960502  0.461520             NaN          0.964770   
10      DoH20  RIPPER  0.954281  0.954245             NaN          0.948266   
11      DoH20     CMR  1.000000  0.999981          100.00          1.000000   
0       KDD99    BRCG  0.973266  0.973377             NaN          0.973340   
1       KDD99  RIPPER  0.998772  0.998386             NaN          0.998772   
2       KDD99     CMR  0.999554  0.999141           69.74          1.000000   
6   UNSW-NB15    BRCG  0.816047  0.767133             NaN          0.816050   
7   UNSW-NB15  RIPPER  0.922908  0.756538             NaN          0.922727   
8   UNSW-NB15     CMR  0.974531  0.824881           

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

#### Avec 1 iteration pour le datasets CIC2017

##### 1. Avec full dataset

In [ ]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {ds: [] for ds in datasets.keys()}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=None,
                dataset_name=name,
                iteration=it+1
            )
            results_all_iterations[name].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = [name, method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds in datasets.keys():
    runs = results_all_iterations.get(ds, [])

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = [ds, method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017.png")
        plt.close()

print(df_results)

results avec 114 mn :  Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Erreur BRCG: Unable to allocate 10.0 GiB for an array with shape (1676837, 803) and data type float64

##### 2. Avec 25k Sample dataset

In [41]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {"CIC2017_25k": []}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=25000,
                dataset_name="CIC2017_25k",
                iteration=it+1
            )
            results_all_iterations["CIC2017_25k"].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = ["CIC2017_25k", method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017_25k.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds, runs in results_all_iterations.items():

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = ["CIC2017_25k", method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017_25k.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017_25k.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017_25k.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017_25k.png")
        plt.close()

print(df_results)

  0%|          | 0/1 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0796
Iteration: 2, Objective: 0.0642


100%|██████████| 1/1 [35:31<00:00, 2131.21s/it]


       Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
0  CIC2017_25k    BRCG  0.938822  0.860065             NaN           0.94336   
1  CIC2017_25k  RIPPER  0.997721  0.995793             NaN           0.99772   
2  CIC2017_25k     CMR  0.998800  0.997473           93.16           1.00000   

   Test_Coverage  Test_Exactitude  Conflit  Non_couverts  F1_rmaj     F1_rr  \
0            NaN         0.883579      NaN           NaN      NaN  0.860065   
1            NaN         0.995794      NaN           NaN      NaN  0.995793   
2          92.38         0.993336     2.51          7.62  0.93864  0.992518   

   Complexité       Temps       RAM_MB  
0         1.0  648.630000   360.710000  
1        34.0   92.070000   637.840000  
2     15953.0  392.772656  1063.738281  


<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

##### 2. Avec 50k Sample dataset

In [40]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {"CIC2017_50k": []}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=50000,
                dataset_name="CIC2017_50k",
                iteration=it+1
            )
            results_all_iterations["CIC2017_50k"].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = ["CIC2017_50k", method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017_50k.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds, runs in results_all_iterations.items():

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = ["CIC2017_50k", method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017_50k.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017_50k.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017_50k.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017_50k.png")
        plt.close()

print(df_results)

  0%|          | 0/1 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0765


100%|██████████| 1/1 [36:20<00:00, 2180.33s/it]


       Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
0  CIC2017_50k    BRCG  0.939141  0.940934             NaN           0.94314   
1  CIC2017_50k  RIPPER  0.998079  0.996849             NaN           0.99808   
2  CIC2017_50k     CMR  0.998559  0.997774           93.22           1.00000   

   Test_Coverage  Test_Exactitude  Conflit  Non_couverts   F1_rmaj     F1_rr  \
0            NaN         0.945157      NaN           NaN       NaN  0.940934   
1            NaN         0.996852      NaN           NaN       NaN  0.996849   
2           92.5         0.993965     2.22           7.5  0.941457  0.993256   

   Complexité       Temps       RAM_MB  
0         1.0  753.950000   446.660000  
1        37.0  155.040000   721.170000  
2     27908.0  200.178662  2064.660156  


<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

##### 3. Avec 100k Sample dataset

In [39]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {"CIC2017_100k": []}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=100000,
                dataset_name="CIC2017_100k",
                iteration=it+1
            )
            results_all_iterations["CIC2017_100k"].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = ["CIC2017_100k", method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017_100k.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds, runs in results_all_iterations.items():

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = ["CIC2017_100k", method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017_100k.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017_100k.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017_100k.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017_100k.png")
        plt.close()

print(df_results)

  0%|          | 0/1 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 1/1 [1:38:15<00:00, 5895.02s/it]


        Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
0  CIC2017_100k    BRCG  0.916561  0.917734             NaN           0.92456   
1  CIC2017_100k  RIPPER  0.997647  0.997151             NaN           0.99765   
2  CIC2017_100k     CMR  0.998399  0.998084           92.85           1.00000   

   Test_Coverage  Test_Exactitude  Conflit  Non_couverts   F1_rmaj     F1_rr  \
0            NaN         0.926335      NaN           NaN       NaN  0.917734   
1            NaN         0.997155      NaN           NaN       NaN  0.997151   
2          92.44         0.994504     1.92          7.56  0.942682  0.993858   

   Complexité        Temps      RAM_MB  
0         1.0  2229.880000   552.10000  
1        38.0   321.450000   951.98000  
2     51950.0   637.777432  1539.96875  


<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

##### 4. Avec 500k Sample dataset

In [ ]:
n_iterations = 1
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {"CIC2017_500k": []}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=500000,
                dataset_name="CIC2017_500k",
                iteration=it+1
            )
            results_all_iterations["CIC2017_500k"].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = ["CIC2017_500k", method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017_500k.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds, runs in results_all_iterations.items():

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = ["CIC2017_500k", method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017_500k.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017_500k.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017_500k.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017_500k.png")
        plt.close()

print(df_results)

  0%|          | 0/1 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001


### 2.3 Évaluation avec 10 iterations avec KDD99, BIG15, UNSW-NB15, DoH20

In [35]:
n_iterations = 10
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        idx = np.random.choice(len(X_train), sample_size, replace=False)
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

    # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None

    return results

datasets = {
    "KDD99": "../Datasets/KDD99/",
    "BIG15": "../Datasets/BIG15/",
    "UNSW-NB15": "../Datasets/UNSW-NB15/processed/",
    "DoH20": "../Datasets/DoH20/"
}

results_all_iterations = {ds: [] for ds in datasets.keys()}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        if name == "UNSW-NB15":
            train = pd.read_csv(path + "le_train.csv")
            test = pd.read_csv(path + "le_test.csv")

            X_train = train.drop(["id", "attack_cat", "label"], axis=1)
            y_train = train["label"]

            X_test = test.drop(["id", "attack_cat", "label"], axis=1)
            y_test = test["label"]
        else:
            X_train = pd.read_csv(path + "X_train.csv")
            X_test = pd.read_csv(path + "X_test.csv")
            y_train = pd.read_csv(path + "y_train.csv").squeeze()
            y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=None,
                dataset_name=name,
                iteration=it+1
            )
            results_all_iterations[name].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = [name, method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations.csv", index=False)
        
        if name == "UNSW-NB15":
            del X_train, X_test, y_train, y_test, train, test
        else:
            del X_train, X_test, y_train, y_test
        gc.collect()


    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds in datasets.keys():
    runs = results_all_iterations.get(ds, [])

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = [ds, method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif.csv", index=False)

if os.path.exists("results/progress_iterations.csv"):
    df_progress = pd.read_csv("results/progress_iterations.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}.png")
        plt.close()

print(df_results)

  0%|          | 0/4 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:02<00:00, 16.57it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 22.13it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 22.52it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:02<00:00, 18.95it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 22.60it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 23.20it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 22.99it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:01<00:00, 22.42it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 39/39 [00:02<00:00, 18.77it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 25%|██▌       | 1/4 [55:13<2:45:41, 3313.80s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:13<00:00,  3.93it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:14<00:00,  3.76it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:13<00:00,  3.92it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:12<00:00,  4.37it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:13<00:00,  4.08it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:14<00:00,  3.71it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:13<00:00,  3.86it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:13<00:00,  3.99it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 54/54 [00:12<00:00,  4.20it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 50%|█████     | 2/4 [2:41:02<2:49:57, 5098.97s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:15<00:00,  2.71it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:15<00:00,  2.73it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:16<00:00,  2.55it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:15<00:00,  2.69it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:14<00:00,  2.99it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:13<00:00,  3.04it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:13<00:00,  3.02it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:13<00:00,  3.01it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 42/42 [00:13<00:00,  3.00it/s]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


 75%|███████▌  | 3/4 [4:38:56<1:40:00, 6000.66s/it]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [02:01<00:00,  4.17s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:47<00:00,  3.71s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:52<00:00,  3.87s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:54<00:00,  3.95s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:49<00:00,  3.78s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:48<00:00,  3.75s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:50<00:00,  3.82s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:53<00:00,  3.90s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 29/29 [01:52<00:00,  3.86s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved


100%|██████████| 4/4 [18:00:52<00:00, 16213.17s/it]


      Dataset Méthode  F1_train   F1_test  Train_Coverage  Train_Exactitude  \
3       BIG15    BRCG  0.981789  0.975395             NaN          0.981736   
4       BIG15  RIPPER  0.984067  0.983954             NaN          0.984032   
5       BIG15     CMR  0.994849  0.993774           95.15          1.000000   
9       DoH20    BRCG  0.960502  0.461520             NaN          0.964770   
10      DoH20  RIPPER  0.954641  0.954434             NaN          0.948615   
11      DoH20     CMR  1.000000  0.999981          100.00          1.000000   
0       KDD99    BRCG  0.973266  0.973377             NaN          0.973340   
1       KDD99  RIPPER  0.998763  0.998351             NaN          0.998763   
2       KDD99     CMR  0.999554  0.999141           69.74          1.000000   
6   UNSW-NB15    BRCG  0.816047  0.767133             NaN          0.816050   
7   UNSW-NB15  RIPPER  0.919536  0.805782             NaN          0.919368   
8   UNSW-NB15     CMR  0.974531  0.824881           

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

### 2.3 Évaluation avec 10 iterations avec CIC2017 sampling 25K

In [ ]:
n_iterations = 10
np.random.seed(42)

os.makedirs("results", exist_ok=True)
os.makedirs("results/distributions", exist_ok=True)

def clean_data(df):
    df = df.copy()
    df = df.replace([np.inf, -np.inf], np.nan)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df.fillna(0)
    return df.astype(np.float32)

def clean_labels(y):
    y = pd.Series(y)
    y = y.replace([np.inf, -np.inf], np.nan)
    y = pd.to_numeric(y, errors="coerce").fillna(0)
    return y.astype(int).values

def test_models_fast(X_train, y_train, X_test, y_test, sample_size=None, dataset_name="", iteration=0):
    results = {}
    process = psutil.Process()

    X_train = clean_data(X_train)
    X_test = clean_data(X_test)

    y_train = clean_labels(y_train)
    y_test = clean_labels(y_test)

    if sample_size is not None and len(X_train) > sample_size:
        rng = np.random.default_rng(42 + iteration)
        idx = rng.choice(len(X_train), sample_size, replace=False) # pour avoir des sample differrent a chaque iteration 
        X_train_sample = X_train.iloc[idx].copy()
        y_train_sample = y_train[idx].copy()
    else:
        X_train_sample = X_train.copy()
        y_train_sample = y_train.copy()

    X_train_sample = X_train_sample.loc[:, X_train_sample.nunique() > 1]
    X_test_aligned = X_test.reindex(columns=X_train_sample.columns, fill_value=0)

    X_train_sample = X_train_sample.astype(np.float32)
    X_test_aligned = X_test_aligned.astype(np.float32)

    y_train_sample = y_train_sample.astype(int).ravel()
    y_test_arr = y_test.astype(int).ravel()

         # -------- BRCG --------
    X_train_brcg = X_train_sample.copy()
    X_test_brcg = X_test_aligned.copy()

    for col in X_train_brcg.columns:
        train_values = set(X_train_brcg[col].unique())
        X_test_brcg[col] = X_test_brcg[col].apply(lambda x: x if x in train_values else 0)

    try:
        (output_brcg, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_brcg,
            X_train_brcg,
            pd.Series(y_train_sample),
            X_test_brcg,
            pd.Series(y_test_arr)
        )

        model_brcg, fb_brcg, columns_brcg, train_time_brcg, metrics_brcg, rules_brcg = output_brcg

        results["BRCG"] = evaluate_brcg(
            model_brcg,
            fb_brcg,
            X_train_brcg,
            y_train_sample,
            X_test_brcg,
            y_test_arr,
            elapsed,
            added_ram,
            columns=columns_brcg
        )

    except Exception as e:
        print(f"Erreur BRCG: {e}")
        results["BRCG"] = None
        
        # -------- RIPPER --------
    try:
        (output_ripper, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            train_ripper,
            X_train_sample,
            pd.Series(y_train_sample),
            X_test_aligned,
            pd.Series(y_test_arr)
        )

        model_ripper, y_pred_ripper, metrics, rules = output_ripper

        results["RIPPER"] = evaluate_ripper(
            model_ripper,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr,
            elapsed,
            added_ram
        )

    except Exception as e:
        print(f"Erreur RIPPER: {e}")
        results["RIPPER"] = None
   
    # -------- CMR --------
    try:
        (output_cmr, elapsed, added_ram, peak_ram_total) = run_with_peak_ram(
            run_cmr,
            X_train_sample,
            y_train_sample,
            X_test_aligned,
            y_test_arr
        )

        results["CMR"] = output_cmr
        results["CMR"]["Temps"] = float(elapsed)
        results["CMR"]["RAM_MB"] = float(added_ram)

    except Exception as e:
        print(f"Erreur CMR: {e}")
        results["CMR"] = None


    return results


datasets = {
    "CIC2017": "../Datasets/CIC2017/"
}

results_all_iterations = {"CIC2017_25k_10iterations": []}
progress_rows = []

columns = [
    "Dataset","Méthode","F1_train","F1_test","Train_Coverage","Train_Exactitude",
    "Test_Coverage","Test_Exactitude","Conflit","Non_couverts","F1_rmaj","F1_rr",
    "Complexité","Temps","RAM_MB","Iteration"
]

for name, path in tqdm(datasets.items()):
    try:
        X_train = pd.read_parquet(path + "X_train.parquet")
        X_test = pd.read_parquet(path + "X_test.parquet")
        y_train = pd.read_csv(path + "y_train.csv").squeeze()
        y_test = pd.read_csv(path + "y_test.csv").squeeze()

        for it in range(n_iterations):
            res = test_models_fast(
                X_train,
                y_train,
                X_test,
                y_test,
                sample_size=25000,
                dataset_name="CIC2017_25k_10iterations",
                iteration=it+1
            )
            results_all_iterations["CIC2017_25k_10iterations"].append(res)

            for method in ["BRCG", "RIPPER", "CMR"]:
                if res.get(method) is not None:
                    row = ["CIC2017_25k_10iterations", method] + list(res[method].values()) + [it+1]
                    progress_rows.append(row)

            df_progress = pd.DataFrame(progress_rows, columns=columns)
            df_progress.to_csv("results/progress_iterations_cic2017_25k_10iterations.csv", index=False)

        del X_train, X_test, y_train, y_test
        gc.collect()

    except Exception as e:
        print(f"\nErreur globale sur le dataset {name}: {e}")
        traceback.print_exc()
        results_all_iterations[name] = [None] * n_iterations
        gc.collect()

final_rows = []

for ds, runs in results_all_iterations.items():

    for method in ["BRCG", "RIPPER", "CMR"]:
        valid_runs = [r[method] for r in runs if r is not None and r.get(method) is not None]
        if valid_runs:
            df_metrics = pd.DataFrame(valid_runs)
            avg = df_metrics.mean(numeric_only=True).to_dict()
            row = ["CIC2017_25k_10iterations", method] + [avg.get(c, None) for c in columns[2:-1]]
            final_rows.append(row)

df_results = pd.DataFrame(final_rows, columns=columns[:-1])

df_results = df_results.sort_values(
    by=["Dataset", "Méthode"],
    key=lambda col: col.map({"BRCG": 0, "RIPPER": 1, "CMR": 2}) if col.name == "Méthode" else col
)

df_results.to_csv("results/tableau_comparatif_cic2017_25k_10iterations.csv", index=False)

if os.path.exists("results/progress_iterations_cic2017_25k_10iterations.csv"):
    df_progress = pd.read_csv("results/progress_iterations_cic2017_25k_10iterations.csv")

    metrics_to_plot = ["F1_test", "Temps", "RAM_MB"]

    for metric in metrics_to_plot:
        plt.figure()
        df_progress.boxplot(column=metric, by=["Dataset", "Méthode"])
        plt.title(metric)
        plt.suptitle("")
        plt.xticks(rotation=45)
        plt.tight_layout()
        plt.savefig(f"results/distributions/{metric}_cic2017_25k_10iterations.png")
        plt.close()

print(df_results)

  0%|          | 0/1 [00:00<?, ?it/s]

Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0786
Iteration: 2, Objective: 0.0619
Iteration: 3, Objective: 0.0619


100%|██████████| 68/68 [01:49<00:00,  1.61s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0762
Iteration: 2, Objective: 0.0609


100%|██████████| 68/68 [03:15<00:00,  2.88s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0780
Iteration: 2, Objective: 0.0615
Iteration: 3, Objective: 0.0602


100%|██████████| 70/70 [05:31<00:00,  4.73s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0795
Iteration: 2, Objective: 0.0650
Iteration: 3, Objective: 0.0617


100%|██████████| 70/70 [07:18<00:00,  6.27s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0735
Iteration: 2, Objective: 0.0574
Iteration: 3, Objective: 0.0571


100%|██████████| 68/68 [03:25<00:00,  3.03s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0769
Iteration: 2, Objective: 0.0616


100%|██████████| 70/70 [06:56<00:00,  5.95s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0821
Iteration: 2, Objective: 0.0651
Iteration: 3, Objective: 0.0651


100%|██████████| 68/68 [06:22<00:00,  5.63s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0780
Iteration: 2, Objective: 0.0612
Iteration: 3, Objective: 0.0602


100%|██████████| 70/70 [09:10<00:00,  7.87s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0778
Iteration: 2, Objective: 0.0640
Iteration: 3, Objective: 0.0616


100%|██████████| 70/70 [06:00<00:00,  5.15s/it]


Learning DNF rule with complexity parameters lambda0=0.001, lambda1=0.001
Initial LP solved
Iteration: 1, Objective: 0.0780
Iteration: 2, Objective: 0.0625
Iteration: 3, Objective: 0.0625


100%|██████████| 1/1 [5:03:08<00:00, 18188.47s/it]


                    Dataset Méthode  F1_train   F1_test  Train_Coverage  \
0  CIC2017_25k_10iterations    BRCG  0.943136  0.941329             NaN   
1  CIC2017_25k_10iterations  RIPPER  0.997675  0.996095             NaN   
2  CIC2017_25k_10iterations     CMR  0.998708  0.997620          92.783   

   Train_Exactitude  Test_Coverage  Test_Exactitude  Conflit  Non_couverts  \
0          0.946952            NaN         0.945348      NaN           NaN   
1          0.997676            NaN         0.996096      NaN           NaN   
2          1.000000         92.263         0.993254    2.488         7.737   

    F1_rmaj     F1_rr  Complexité       Temps      RAM_MB  
0       NaN  0.941329         1.0  643.620000   545.99500  
1       NaN  0.996095        30.7   85.229000   625.82000  
2  0.938002  0.992452     15791.3  369.752181  6441.56875  


<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

<Figure size 640x480 with 0 Axes>

: 

In [10]:
import sys
import sklearn
import numpy
import pandas

print("Python:", sys.version)
print("sklearn:", sklearn.__version__)
print("numpy:", numpy.__version__)
print("pandas:", pandas.__version__)

Python: 3.9.25 (main, Nov  3 2025, 22:44:01) [MSC v.1929 64 bit (AMD64)]
sklearn: 1.6.1
numpy: 1.23.5
pandas: 1.5.3


In [11]:
import platform
import psutil

print("OS :", platform.system(), platform.release())
print("CPU :", platform.processor())
print("RAM :", round(psutil.virtual_memory().total / (1024**3), 2), "GB")

OS : Windows 10
CPU : Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
RAM : 15.41 GB


In [16]:
import pandas as pd
import os

datasets = {
    "KDD99": "../Datasets/KDD99/",
    "BIG15": "../Datasets/BIG15/",
    "DoH20": "../Datasets/DoH20/",
    "UNSW-NB15": "../Datasets/UNSW-NB15/processed/",
    "CIC2017": "../Datasets/CIC2017/"
}

results = []

for name, path in datasets.items():
    try:
        if name == "CIC2017":
            X_train = pd.read_parquet(path + "X_train.parquet")
            X_test = pd.read_parquet(path + "X_test.parquet")

        elif name == "UNSW-NB15":
            train = pd.read_csv(path + "le_train.csv")
            test = pd.read_csv(path + "le_test.csv")

            X_train = train.drop(["id", "attack_cat", "label"], axis=1)
            X_test = test.drop(["id", "attack_cat", "label"], axis=1)

        else:
            X_train = pd.read_csv(path + "X_train.csv")
            X_test = pd.read_csv(path + "X_test.csv")

        results.append({
            "Dataset": name,
            "Train Samples": X_train.shape[0],
            "Test Samples": X_test.shape[0],
            "Features": X_train.shape[1]
        })

    except Exception as e:
        print(f"Erreur avec {name}: {e}")

df_sizes = pd.DataFrame(results)
print(df_sizes)

     Dataset  Train Samples  Test Samples  Features
0      KDD99         116468         29117        39
1      BIG15         110437         27610        54
2      DoH20         214928         53733        29
3  UNSW-NB15          82332        175341        42
4    CIC2017        2017884        504472        78


Les expériences ont été réalisées dans l’environnement technique suivant :

Système d’exploitation : Windows 10
Processeur : Intel64 Family 6 Model 78 (GenuineIntel)
Mémoire RAM : 15.41 Go
Langage de programmation : Python 3.9.25 (64 bits)

Bibliothèques utilisées :

scikit-learn : 1.6.1
NumPy : 1.23.5
pandas : 1.5.3

Les expériences ont été exécutées en local sur cette machine.

Resume tableau 8 : Les résultats montrent qu'augmenter le paramètre de support minimal (s_min) réduit le nombre de règles extraites, ce qui diminue les taux de conflit mais aussi la couverture. 

Si les performances peuvent s'améliorer légèrement grâce à la réduction du bruit dans certains jeux de données, elles peuvent également se dégrader lorsque des motifs importants sont supprimés. Notamment, s_min = 1 et s_min = 3 produisent des résultats identiques, indiquant que toutes les règles extraites bénéficient déjà d'un support suffisant. Globalement, s_min = 3 offre un bon compromis en termes de couverture.…